recursion/gnark-ffi/src/groth16.rs

```diff
@@ -26,9 +33,73 @@ pub struct Groth16Proof {
 }
 
 impl Groth16Prover {
-    /// Creates a new verifier.
+    /// Starts up the Gnark server using Groth16 on the given port and waits for it to be ready.
     pub fn new() -> Self {
-        Groth16Prover
+        let manifest_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+        let gnark_dir = manifest_dir.join("../gnark");
+        let port = env::var("HOST_PORT").unwrap_or_else(|_| generate_random_port().to_string());
+        let port_clone = port.clone();
+
+        // Spawn a thread to run the command
+        // TODO: version by commit hash instead of by incrementing
+        thread::spawn(move || {
+            let mut child = Command::new("go")
+                .args([
+                    "run",
+                    "main.go",
+                    "serve",
+                    "--type",
+                    "groth16",
+                    "--version",
+                    "1",
+                    "--port",
+                    &port,
+                ])
+                .current_dir(gnark_dir)
+                .stderr(Stdio::inherit())
+                .stdout(Stdio::inherit())
+                .stdin(Stdio::inherit())
+                .spawn()
+                .unwrap();
+
+            let exit_status = child.wait().unwrap();
+
+            if !exit_status.success() {
+                panic!("Gnark server exited with an error: {:?}", exit_status);
+            }
+        });
+
+        let prover = Self { port: port_clone };
+
+        prover.wait_for_healthy_server().unwrap();
+
+        prover
+    }
+
+    /// Checks if the server is ready to accept requests.
+    fn wait_for_healthy_server(&self) -> Result<(), Box<dyn std::error::Error>> {
```

Comment:
I'm still kinda bullish doing this in prove() instead of new(), because then it's really easy for this to just be async work. For example:
```
    let wrapper_client = Groth16Prover::new();
    let sp1_prover = SP1Prover::new();
    // plus some other stuff
    // plus core proof
    // then finally, groth16 prove()
    ```

Reply:
maybe start it in new but expose wait

Reply:
maybe its fine for now dont want to expose all the async